src/main/scala/viper/silver/plugin/standard/inline/PredicateExpansion.scala

```diff
@@ -0,0 +1,24 @@
+package viper.silver.plugin.standard.inline
+
+import viper.silver.ast.utility.ViperStrategy
+import viper.silver.ast.{Exp, FieldAccessPredicate, PredicateAccessPredicate}
+import viper.silver.ast.utility.rewriter.Traverse
+
+trait PredicateExpansion {
+
+  def propagatePermission(maybeBody: Option[Exp], originalPerm: Exp): Option[Exp] =
```

Comment:
For a given predicate body, we thread the original permission value in. If we find any accesses, e.g. `FieldAccess` or a `PredicateAccess`, we replace its permission value with the original one.